pwndbg/gdblib/config.py

```diff
@@ -0,0 +1,101 @@
+"""
+Dynamic configuration system for pwndbg, using GDB's built-in Parameter
+mechanism.
+
+To create a new pwndbg configuration point, create an instance of
+``pwndbg.lib.config.Parameter``.
+
+Parameters should be declared in the module in which they are primarily
+used, or in this module for general-purpose parameters.
+
+All pwndbg Parameter types are accessible via property access on this
+module, for example:
+
+    >>> pwndbg.gdblib.config.add_param('example-value', 7, 'an example')
+    >>> int(pwndbg.gdblib.config.example_value)
+    7
+"""
+import collections
+
+import gdb
+
+import pwndbg.decorators
+import pwndbg.lib.config
+
+config = pwndbg.lib.config.Config()
+
+# TODO: I don't think this needs to be an OrderedDict
+PARAM_CLASSES = collections.OrderedDict()
+
+# The value is a plain boolean.
+# The Python boolean values, True and False are the only valid values.
+PARAM_CLASSES[bool] = gdb.PARAM_BOOLEAN
+
+# The value is an integer.
+# This is like PARAM_INTEGER, except 0 is interpreted as itself.
+PARAM_CLASSES[int] = gdb.PARAM_ZINTEGER
+
+# The value is a string.
+# When the user modifies the string, any escape sequences,
+# such as ‘\t’, ‘\f’, and octal escapes, are translated into
+# corresponding characters and encoded into the current host charset.
+PARAM_CLASSES[str] = gdb.PARAM_STRING
+
+
+# See this for details about the API of `gdb.Parameter`:
+# https://sourceware.org/gdb/onlinedocs/gdb/Parameters-In-Python.html
+class Parameter(gdb.Parameter):
+    def __init__(self, param: pwndbg.lib.config.Parameter):
+        self.param = param
+
+        # `set_doc` and `show_doc` must be set before `gdb.Parameter.__init__`.
+        # They will be used for `help set <param>` and `help show <param>`,
+        # respectively
+        self.set_doc = "Set " + self.param.docstring
+        self.show_doc = "Show " + self.param.docstring
```

Comment:
Just noting here: we should probably have a way to overwrite set/show docs for some of the config values

Reply:
We're passing in the docstring when we define the parameters, so when would we want to overwrite that?

Reply:
We would probably have to change the interface, or, give an option to either provide docstring or set+show strings?

My point is, we could display some more information in e.g. those parameters:
```
pwndbg> show resolve-heap-via-heuristic
Sets Resolve missing heap related symbols via heuristics (currently: False)

pwndbg> show context-sections
Sets which context sections are displayed (controls order) (currently: 'regs disasm code ghidra stack backtrace expressions')

pwndbg> show debug-events
Sets display internal event debugging info (currently: False)
```

Also, should we maybe mark that those are comming from pwndbg? Ah, I think we wanted to do it with namespaces :+1: